File: lib/sidereon/drag.ex
defmodule Sidereon.Drag do
  @moduledoc """
  Atmospheric drag parameters, acceleration, and decay estimates.
  """

  alias Sidereon.Astro.Relative.State
  alias Sidereon.NIF

  defmodule SpaceWeather do
    @moduledoc """
    Space-weather inputs for the core drag model.
    """
    @enforce_keys [:f107, :f107a, :ap]
    defstruct [:f107, :f107a, :ap]
    @type t :: %__MODULE__{f107: float(), f107a: float(), ap: float()}
  end

  defmodule Parameters do
    @moduledoc """
    Ballistic drag settings passed to propagation and drag-force calculations.
    """
    @enforce_keys [:bc_factor_m2_kg, :space_weather, :cutoff_altitude_km]
    defstruct [:bc_factor_m2_kg, :space_weather, :cutoff_altitude_km]

    @type t :: %__MODULE__{
            bc_factor_m2_kg: float(),
            space_weather: SpaceWeather.t(),
            cutoff_altitude_km: float()
          }
  end

  defmodule DecayEstimate do
    @moduledoc """
    Estimated decay time and reentry state.
    """
    @enforce_keys [:time_to_decay_s, :reentry_state, :reentry_altitude_km]
    defstruct [:time_to_decay_s, :reentry_state, :reentry_altitude_km]

    @type t :: %__MODULE__{
            time_to_decay_s: float(),
            reentry_state: State.t(),
            reentry_altitude_km: float()
          }
  end

  @default_cutoff_altitude_km 100.0

  @spec default_space_weather() :: SpaceWeather.t()
  def default_space_weather do
    fields = NIF.drag_space_weather_default()
    %SpaceWeather{f107: fields.f107, f107a: fields.f107a, ap: fields.ap}
  end

  @spec from_area_mass(number(), number(), number(), keyword()) :: {:ok, Parameters.t()} | {:error, atom()}
  def from_area_mass(cd, area_m2, mass_kg, opts \\ []) do
    call_params(:drag_parameters_from_area_mass, [
      cd / 1.0,
      area_m2 / 1.0,
      mass_kg / 1.0,
      space_weather_map(Keyword.get(opts, :space_weather, default_space_weather())),
      Keyword.get(opts, :cutoff_altitude_km, @default_cutoff_altitude_km) / 1.0
    ])
  end

  @spec from_bc_factor(number(), keyword()) :: {:ok, Parameters.t()} | {:error, atom()}
  def from_bc_factor(bc_factor_m2_kg, opts \\ []) do
    call_params(:drag_parameters_from_bc_factor, [
      bc_factor_m2_kg / 1.0,
      space_weather_map(Keyword.get(opts, :space_weather, default_space_weather())),
      Keyword.get(opts, :cutoff_altitude_km, @default_cutoff_altitude_km) / 1.0
    ])
  end

  @spec from_ballistic_coefficient(number(), keyword()) :: {:ok, Parameters.t()} | {:error, atom()}
  def from_ballistic_coefficient(bc_kg_m2, opts \\ []) do
    call_params(:drag_parameters_from_ballistic_coefficient, [
      bc_kg_m2 / 1.0,
      space_weather_map(Keyword.get(opts, :space_weather, default_space_weather())),
      Keyword.get(opts, :cutoff_altitude_km, @default_cutoff_altitude_km) / 1.0
    ])
  end

  @spec acceleration(Parameters.t(), State.t()) :: {:ok, {float(), float(), float()}} | {:error, atom()}
  def acceleration(%Parameters{} = params, %State{} = state) do
    NIF.drag_force_acceleration(params_map(params), Map.from_struct(state))
  rescue
    e in ErlangError -> {:error, e.original}
  end

  @spec estimate_decay(State.t(), Parameters.t(), keyword()) :: {:ok, DecayEstimate.t()} | {:error, term()}
  def estimate_decay(%State{} = state, %Parameters{} = params, opts \\ []) do
    case NIF.drag_estimate_decay(
           Map.from_struct(state),
           params_map(params),
           Keyword.get(opts, :force_model, :twobody) |> Atom.to_string(),
           Keyword.get(opts, :abs_tol, 1.0e-9) / 1.0,
           Keyword.get(opts, :rel_tol, 1.0e-12) / 1.0,
           Keyword.get(opts, :reentry_altitude_km, @default_cutoff_altitude_km) / 1.0,
           Keyword.get(opts, :scan_step_s, 60.0) / 1.0,
           Keyword.get(opts, :crossing_tolerance_s, 1.0) / 1.0,
           Keyword.get(opts, :max_duration_s, 12_000_000.0) / 1.0,
           Keyword.get(opts, :max_scan_samples, 200_000)
         ) do
      {:ok, fields} -> {:ok, to_decay(fields)}
      {:error, reason} -> {:error, reason}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  def from_area_mass!(cd, area_m2, mass_kg, opts \\ []), do: bang(from_area_mass(cd, area_m2, mass_kg, opts))
  def from_bc_factor!(bc_factor_m2_kg, opts \\ []), do: bang(from_bc_factor(bc_factor_m2_kg, opts))
  def from_ballistic_coefficient!(bc_kg_m2, opts \\ []), do: bang(from_ballistic_coefficient(bc_kg_m2, opts))
  def acceleration!(params, state), do: bang(acceleration(params, state))
  def estimate_decay!(state, params, opts \\ []), do: bang(estimate_decay(state, params, opts))

  defp call_params(fun, args) do
    case apply(NIF, fun, args) do
      {:ok, fields} -> {:ok, to_params(fields)}
      {:error, reason} -> {:error, reason}
    end
  rescue
    e in ErlangError -> {:error, e.original}
  end

  defp to_params(fields) do
    %Parameters{
      bc_factor_m2_kg: fields.bc_factor_m2_kg,
      space_weather: %SpaceWeather{f107: fields.f107, f107a: fields.f107a, ap: fields.ap},
      cutoff_altitude_km: fields.cutoff_altitude_km
    }
  end

  defp to_decay(fields) do
    state = fields.reentry_state

    %DecayEstimate{
      time_to_decay_s: fields.time_to_decay_s,
      reentry_altitude_km: fields.reentry_altitude_km,
      reentry_state: %State{
        epoch_tdb_seconds: state.epoch_tdb_seconds,
        position_km: state.position_km,
        velocity_km_s: state.velocity_km_s
      }
    }
  end

  defp params_map(%Parameters{} = params) do
    %{
      bc_factor_m2_kg: params.bc_factor_m2_kg,
      f107: params.space_weather.f107,
      f107a: params.space_weather.f107a,
      ap: params.space_weather.ap,
      cutoff_altitude_km: params.cutoff_altitude_km
    }
  end

  defp space_weather_map(%SpaceWeather{} = sw), do: Map.from_struct(sw)
  defp bang({:ok, value}), do: value
  defp bang({:error, reason}), do: raise(ArgumentError, "drag calculation failed: #{inspect(reason)}")
end

File: native/sidereon_nif/src/drag.rs
use rustler::{Encoder, Env, Error, NifResult, Term};
use sidereon_core::astro::forces::r#trait::ForceModel;
use sidereon_core::astro::forces::{DragParameters, SpaceWeather};
use sidereon_core::astro::propagator::decay::{estimate_decay, DecayConfig};
use sidereon_core::astro::propagator::{
    IntegratorOptions, PropagationContext, PropagationForceModel,
};
use sidereon_core::astro::state::CartesianState;

use crate::errors;

type Vec3 = (f64, f64, f64);

mod atoms {
    rustler::atoms! {
        ok,
        error,
        invalid_input,
        no_decay_within_horizon,
        scan_budget_exhausted,
        propagation_failed
    }
}

#[derive(Debug, Clone, rustler::NifMap)]
pub(crate) struct DragParametersTerm {
    bc_factor_m2_kg: f64,
    f107: f64,
    f107a: f64,
    ap: f64,
    cutoff_altitude_km: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct SpaceWeatherTerm {
    f107: f64,
    f107a: f64,
    ap: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct CartesianStateTerm {
    epoch_tdb_seconds: f64,
    position_km: Vec3,
    velocity_km_s: Vec3,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct DecayEstimateTerm {
    time_to_decay_s: f64,
    reentry_state: CartesianStateTerm,
    reentry_altitude_km: f64,
}

pub(crate) fn decode_drag_parameters(term: Term<'_>) -> NifResult<DragParameters> {
    let decoded: DragParametersTerm = term.decode()?;
    DragParameters::from_bc_factor_m2_kg(
        decoded.bc_factor_m2_kg,
        SpaceWeather {
            f107: decoded.f107,
            f107a: decoded.f107a,
            ap: decoded.ap,
        },
        decoded.cutoff_altitude_km,
    )
    .map_err(errors::invalid_input)
}

fn params_to_term(params: DragParameters) -> DragParametersTerm {
    let sw = params.space_weather();
    DragParametersTerm {
        bc_factor_m2_kg: params.bc_factor_m2_kg(),
        f107: sw.f107,
        f107a: sw.f107a,
        ap: sw.ap,
        cutoff_altitude_km: params.cutoff_altitude_km(),
    }
}

fn decode_space_weather(term: SpaceWeatherTerm) -> SpaceWeather {
    SpaceWeather {
        f107: term.f107,
        f107a: term.f107a,
        ap: term.ap,
    }
}

fn state_from_term(term: CartesianStateTerm) -> CartesianState {
    CartesianState::new(
        term.epoch_tdb_seconds,
        [term.position_km.0, term.position_km.1, term.position_km.2],
        [
            term.velocity_km_s.0,
            term.velocity_km_s.1,
            term.velocity_km_s.2,
        ],
    )
}

fn state_to_term(state: CartesianState) -> CartesianStateTerm {
    let position = state.position_array();
    let velocity = state.velocity_array();
    CartesianStateTerm {
        epoch_tdb_seconds: state.epoch_tdb_seconds,
        position_km: (position[0], position[1], position[2]),
        velocity_km_s: (velocity[0], velocity[1], velocity[2]),
    }
}

fn force_model(name: &str) -> NifResult<PropagationForceModel> {
    Ok(match name {
        "twobody" => PropagationForceModel::TwoBody,
        "j2" => PropagationForceModel::TwoBodyJ2,
        _ => return Err(Error::Term(Box::new("unknown force model"))),
    })
}

#[rustler::nif]
fn drag_space_weather_default() -> SpaceWeatherTerm {
    let sw = SpaceWeather::default();
    SpaceWeatherTerm {
        f107: sw.f107,
        f107a: sw.f107a,
        ap: sw.ap,
    }
}

#[rustler::nif]
fn drag_parameters_from_area_mass<'a>(
    env: Env<'a>,
    cd: f64,
    area_m2: f64,
    mass_kg: f64,
    space_weather: SpaceWeatherTerm,
    cutoff_altitude_km: f64,
) -> Term<'a> {
    match DragParameters::from_area_mass(
        cd,
        area_m2,
        mass_kg,
        decode_space_weather(space_weather),
        cutoff_altitude_km,
    ) {
        Ok(params) => (atoms::ok(), params_to_term(params)).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn drag_parameters_from_bc_factor<'a>(
    env: Env<'a>,
    bc_factor_m2_kg: f64,
    space_weather: SpaceWeatherTerm,
    cutoff_altitude_km: f64,
) -> Term<'a> {
    match DragParameters::from_bc_factor_m2_kg(
        bc_factor_m2_kg,
        decode_space_weather(space_weather),
        cutoff_altitude_km,
    ) {
        Ok(params) => (atoms::ok(), params_to_term(params)).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn drag_parameters_from_ballistic_coefficient<'a>(
    env: Env<'a>,
    bc_kg_m2: f64,
    space_weather: SpaceWeatherTerm,
    cutoff_altitude_km: f64,
) -> Term<'a> {
    match DragParameters::from_ballistic_coefficient(
        bc_kg_m2,
        decode_space_weather(space_weather),
        cutoff_altitude_km,
    ) {
        Ok(params) => (atoms::ok(), params_to_term(params)).encode(env),
        Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
    }
}

#[rustler::nif]
fn drag_force_acceleration<'a>(
    env: Env<'a>,
    params: Term<'a>,
    state: CartesianStateTerm,
) -> NifResult<Term<'a>> {
    let force = decode_drag_parameters(params)?.to_force();
    let state = state_from_term(state);
    Ok(
        match force.acceleration(&state, &PropagationContext::default()) {
            Ok(a) => (atoms::ok(), (a.x, a.y, a.z)).encode(env),
            Err(_) => (atoms::error(), atoms::invalid_input()).encode(env),
        },
    )
}

#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn drag_estimate_decay<'a>(
    env: Env<'a>,
    state: CartesianStateTerm,
    params: Term<'a>,
    force_model_name: String,
    abs_tol: f64,
    rel_tol: f64,
    reentry_altitude_km: f64,
    scan_step_s: f64,
    crossing_tolerance_s: f64,
    max_duration_s: f64,
    max_scan_samples: u32,
) -> NifResult<Term<'a>> {
    let mut config = DecayConfig::new(decode_drag_parameters(params)?)
        .with_force_model(force_model(&force_model_name)?)
        .with_options(IntegratorOptions {
            abs_tol,
            rel_tol,
            ..IntegratorOptions::default()
        })
        .with_reentry_altitude_km(reentry_altitude_km)
        .with_scan_step_s(scan_step_s)
        .with_crossing_tolerance_s(crossing_tolerance_s)
        .with_max_duration_s(max_duration_s)
        .with_max_scan_samples(max_scan_samples);
    config.mu_km3_s2 = None;

    Ok(match estimate_decay(state_from_term(state), &config) {
        Ok(estimate) => (
            atoms::ok(),
            DecayEstimateTerm {
                time_to_decay_s: estimate.time_to_decay_s,
                reentry_state: state_to_term(estimate.reentry_state),
                reentry_altitude_km: estimate.reentry_altitude_km,
            },
        )
            .encode(env),
        Err(error) => match error {
            sidereon_core::astro::propagator::decay::DecayError::NoDecayWithinHorizon {
                horizon_s,
            } => (
                atoms::error(),
                (atoms::no_decay_within_horizon(), horizon_s),
            )
                .encode(env),
            sidereon_core::astro::propagator::decay::DecayError::ScanBudgetExhausted {
                scanned_s,
                samples,
            } => (
                atoms::error(),
                (atoms::scan_budget_exhausted(), scanned_s, samples),
            )
                .encode(env),
            sidereon_core::astro::propagator::decay::DecayError::Propagation(_) => {
                (atoms::error(), atoms::propagation_failed()).encode(env)
            }
            sidereon_core::astro::propagator::decay::DecayError::InvalidConfig(_) => {
                (atoms::error(), atoms::invalid_input()).encode(env)
            }
        },
    })
}
